File: youtube/lat8-heart/try2.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns 
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_selection import RFE
import statsmodels.api as sm 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold, cross_val_score
from sklearn.neural_network import MLPClassifier, MLPRegressor


heart = pd.read_csv('ML/youtube/lat8-heart/heart.csv')
print(heart)
print(heart.info())
print(heart.describe())
print(heart.isnull().sum())

for col in heart:
    print(col)
    print(heart[col].unique())
    print('\n\n')
    
# fig, ax = plt.subplots(3, 5, figsize=(15, 10))
# sns.scatterplot(x=heart.index, y=heart['age'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=heart.index, y=heart['sex'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=heart.index, y=heart['cp'], ax=ax[0, 2], color='orange')
# sns.scatterplot(x=heart.index, y=heart['trestbps'], ax=ax[0, 3], color='red')
# sns.scatterplot(x=heart.index, y=heart['chol'], ax=ax[0, 4], color='blue')
# sns.scatterplot(x=heart.index, y=heart['fbs'], ax=ax[1, 0], color='green')
# sns.scatterplot(x=heart.index, y=heart['restecg'], ax=ax[1, 1], color='orange')
# sns.scatterplot(x=heart.index, y=heart['thalach'], ax=ax[1, 2], color='red')
# sns.scatterplot(x=heart.index, y=heart['exang'], ax=ax[1, 3], color='blue')
# sns.scatterplot(x=heart.index, y=heart['oldpeak'], ax=ax[1, 4], color='green')
# sns.scatterplot(x=heart.index, y=heart['slope'], ax=ax[2, 0], color='orange')
# sns.scatterplot(x=heart.index, y=heart['ca'], ax=ax[2, 1], color='red')
# sns.scatterplot(x=heart.index, y=heart['thal'], ax=ax[2, 2], color='red')
# sns.scatterplot(x=heart.index, y=heart['target'], ax=ax[2, 3], color='red')
# plt.tight_layout()
# plt.show()

q1 = heart['age'].quantile(0.25)
q3 = heart['age'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['age'] >= q1 - 1.5*iqr) & (heart['age'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['sex'], 25)
q3 = np.percentile(heart['sex'], 75)
iqr = q3 - q1
heart = heart[(heart['sex'] >= q1 - 1.5*iqr) & (heart['sex'] <= q3 + 1.5*iqr )]

q1 = heart['cp'].quantile(0.25)
q3 = heart['cp'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['cp'] >= q1 - 1.5*iqr) & (heart['cp'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['trestbps'], 25)
q3 = np.percentile(heart['trestbps'], 75)
iqr = q3 - q1
heart = heart[(heart['trestbps'] >= q1 - 1.5*iqr) & (heart['trestbps'] <= q3 + 1.5*iqr )]

q1 = heart['chol'].quantile(0.25)
q3 = heart['chol'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['chol'] >= q1 - 1.5*iqr) & (heart['chol'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['fbs'], 25)
q3 = np.percentile(heart['fbs'], 75)
iqr = q3 - q1
heart = heart[(heart['fbs'] >= q1 - 1.5*iqr) & (heart['fbs'] <= q3 + 1.5*iqr )]

q1 = heart['restecg'].quantile(0.25)
q3 = heart['restecg'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['restecg'] >= q1 - 1.5*iqr) & (heart['restecg'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['thalach'], 25)
q3 = np.percentile(heart['thalach'], 75)
iqr = q3 - q1
heart = heart[(heart['thalach'] >= q1 - 1.5*iqr) & (heart['thalach'] <= q3 + 1.5*iqr )]

q1 = heart['exang'].quantile(0.25)
q3 = heart['exang'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['exang'] >= q1 - 1.5*iqr) & (heart['exang'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['oldpeak'], 25)
q3 = np.percentile(heart['oldpeak'], 75)
iqr = q3 - q1
heart = heart[(heart['oldpeak'] >= q1 - 1.5*iqr) & (heart['oldpeak'] <= q3 + 1.5*iqr )]

q1 = heart['slope'].quantile(0.25)
q3 = heart['slope'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['slope'] >= q1 - 1.5*iqr) & (heart['slope'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['ca'], 25)
q3 = np.percentile(heart['ca'], 75)
iqr = q3 - q1
heart = heart[(heart['ca'] >= q1 - 1.5*iqr) & (heart['ca'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['thal'], 25)
q3 = np.percentile(heart['thal'], 75)
iqr = q3 - q1
heart = heart[(heart['thal'] >= q1 - 1.5*iqr) & (heart['thal'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['target'], 25)
q3 = np.percentile(heart['target'], 75)
iqr = q3 - q1
heart = heart[(heart['target'] >= q1 - 1.5*iqr) & (heart['target'] <= q3 + 1.5*iqr )]

# fig, ax = plt.subplots(3, 5, figsize=(15, 10))
# sns.scatterplot(x=heart.index, y=heart['age'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=heart.index, y=heart['sex'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=heart.index, y=heart['cp'], ax=ax[0, 2], color='orange')
# sns.scatterplot(x=heart.index, y=heart['trestbps'], ax=ax[0, 3], color='red')
# sns.scatterplot(x=heart.index, y=heart['chol'], ax=ax[0, 4], color='blue')
# sns.scatterplot(x=heart.index, y=heart['fbs'], ax=ax[1, 0], color='green')
# sns.scatterplot(x=heart.index, y=heart['restecg'], ax=ax[1, 1], color='orange')
# sns.scatterplot(x=heart.index, y=heart['thalach'], ax=ax[1, 2], color='red')
# sns.scatterplot(x=heart.index, y=heart['exang'], ax=ax[1, 3], color='blue')
# sns.scatterplot(x=heart.index, y=heart['oldpeak'], ax=ax[1, 4], color='green')
# sns.scatterplot(x=heart.index, y=heart['slope'], ax=ax[2, 0], color='orange')
# sns.scatterplot(x=heart.index, y=heart['ca'], ax=ax[2, 1], color='red')
# sns.scatterplot(x=heart.index, y=heart['thal'], ax=ax[2, 2], color='red')
# sns.scatterplot(x=heart.index, y=heart['target'], ax=ax[2, 3], color='red')
# plt.tight_layout()
# plt.show()

numvars = ['age', 'trestbps', 'chol', 'thalach', 'oldpeak']
scaler = MinMaxScaler()
heart[numvars] = scaler.fit_transform(heart[numvars])

print(heart)


df_train, df_test = train_test_split(heart, train_size = 0.7, test_size=0.3, random_state=100)
print(df_train.head())
print(df_test.head())

corr = df_train.corr()
print(corr)
# plt.figure(figsize=(16, 10))
# sns.heatmap(corr, annot = True, cmap="YlGnBu")
# plt.show()


x_train = df_train
y_train = df_train.pop('target')

k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelRF  = RandomForestClassifier()
scoring = 'accuracy'
score = cross_val_score(modelRF, x_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelRF.fit(x_train, y_train)



# sebenarnya tidak dibutuhkan karena jumlah kolomnya sendiri hanya 4
rfe = RFE(estimator=modelRF, n_features_to_select=5)
rfe = rfe.fit(x_train, y_train)
print(rfe.support_)
print(rfe.ranking_)

hasil = pd.DataFrame({
    'features': x_train.columns,
    'support': rfe.support_,
    'rank': rfe.ranking_,
})
print(hasil)

support = x_train.columns[rfe.support_]
not_support = x_train.columns[~rfe.support_]
print(support)
print(not_support)

x_train_rfe = x_train[support]
# x_train_rfe = sm.add_constant(x_train_rfe)
print(x_train_rfe.head())
print(x_train_rfe.shape)


k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelRF  = RandomForestClassifier()
scoring = 'accuracy'
score = cross_val_score(modelRF, x_train_rfe, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelRF.fit(x_train_rfe, y_train)



vif = pd.DataFrame()
x = x_train_rfe
print(x.shape)

vif['features'] = x.columns
vif['VIF'] = [variance_inflation_factor(x.values, i) for i in range(x.shape[1])]
vif = vif.sort_values('VIF', ascending=False)
print(vif)

vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values('VIF', ascending=False)
print(vif)

y_pred = modelRF.predict(x_train_rfe)

df = pd.DataFrame({
    'y_train': y_train,
    'y_pred': y_pred,
})
print(df.head())
print(df[df['y_train'] != df['y_pred']].head())

akurasi = accuracy_score(y_train, y_pred)
print(akurasi)
cm = confusion_matrix(y_train, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()


x_test = df_test
x_test_rfe = x_test[support]
y_test = df_test.pop('target')
y_pred = modelRF.predict(x_test_rfe)

df = pd.DataFrame({
    'y_test': y_test,
    'y_pred': y_pred,
})
print(df.head())
print(df[df['y_test'] != df['y_pred']])

akurasi = accuracy_score(y_test, y_pred)
print(akurasi)
cm = confusion_matrix(y_test, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

File: youtube/lat8-heart/try2.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(corrplot)
library(caTools)
library(dplyr)
library(fastDummies)
library(zoo)
library(pROC)
library(ROCR)
library(leaps)
library(car)
library(tidyr)
library(plyr)
library(lubridate)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees




df <- read.csv("ML/youtube/lat8-heart/heart.csv")
print(head(df))
print(dim(df))
print(str(df))
print(summary(df))
print(colSums(is.na(df)))
print(sapply(df, class))
print(sapply(df, function(x) length(unique(x[!is.na(x)]))))


for (col in colnames(df)){
    print(col)
    print(unique(df[[col]]))
    cat('\n\n')
}


p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = age), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = sex), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = cp), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = trestbps), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = chol), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fbs), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = restecg), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = thalach), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = exang), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = oldpeak), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = slope), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = ca), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p13 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = thal), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p14 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = target), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # Menggabungkan plot menggunakan patchwork
plot <- p1+p2+p3+p4+p5+p6+p7+p8+p9+p10+p11+p12+p13+p14
# plot <- (p1 | p2) / (p3 | p4)
print(plot)



q1 <- quantile(df$age, 0.25)
q3 <- quantile(df$age, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, age >= q1 - 1.5 * iqr & age <= q3 + 1.5 * iqr)

q1 <- quantile(df$sex, 0.25)
q3 <- quantile(df$sex, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, sex >= q1 - 1.5 * iqr & sex <= q3 + 1.5 * iqr)

q1 <- quantile(df$cp , 0.25)
q3 <- quantile(df$cp , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, cp  >= q1 - 1.5 * iqr & cp  <= q3 + 1.5 * iqr)

q1 <- quantile(df$trestbps , 0.25)
q3 <- quantile(df$trestbps , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, trestbps  >= q1 - 1.5 * iqr & trestbps  <= q3 + 1.5 * iqr)

q1 <- quantile(df$chol  , 0.25)
q3 <- quantile(df$chol  , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, chol   >= q1 - 1.5 * iqr & chol   <= q3 + 1.5 * iqr)

q1 <- quantile(df$fbs , 0.25)
q3 <- quantile(df$fbs , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, fbs  >= q1 - 1.5 * iqr & fbs  <= q3 + 1.5 * iqr)

q1 <- quantile(df$restecg , 0.25)
q3 <- quantile(df$restecg , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, restecg  >= q1 - 1.5 * iqr & restecg  <= q3 + 1.5 * iqr)

q1 <- quantile(df$thalach, 0.25)
q3 <- quantile(df$thalach, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, thalach >= q1 - 1.5 * iqr & thalach <= q3 + 1.5 * iqr)

q1 <- quantile(df$exang , 0.25)
q3 <- quantile(df$exang , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, exang  >= q1 - 1.5 * iqr & exang  <= q3 + 1.5 * iqr)

q1 <- quantile(df$oldpeak, 0.25)
q3 <- quantile(df$oldpeak, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, oldpeak >= q1 - 1.5 * iqr & oldpeak <= q3 + 1.5 * iqr)

q1 <- quantile(df$slope , 0.25)
q3 <- quantile(df$slope , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, slope  >= q1 - 1.5 * iqr & slope  <= q3 + 1.5 * iqr)

q1 <- quantile(df$ca, 0.25)
q3 <- quantile(df$ca, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, ca >= q1 - 1.5 * iqr & ca <= q3 + 1.5 * iqr)

q1 <- quantile(df$thal , 0.25)
q3 <- quantile(df$thal , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, thal  >= q1 - 1.5 * iqr & thal  <= q3 + 1.5 * iqr)

q1 <- quantile(df$target , 0.25)
q3 <- quantile(df$target , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, target  >= q1 - 1.5 * iqr & target  <= q3 + 1.5 * iqr)



p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = age), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = sex), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = cp), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = trestbps), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = chol), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fbs), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = restecg), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = thalach), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = exang), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = oldpeak), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = slope), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = ca), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p13 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = thal), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p14 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = target), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # Menggabungkan plot menggunakan patchwork
plot <- p1+p2+p3+p4+p5+p6+p7+p8+p9+p10+p11+p12+p13+p14
# plot <- (p1 | p2) / (p3 | p4)
print(plot)


print(head(df))

numvars <- c('age', 'trestbps', 'chol', 'thalach', 'oldpeak')
scaler <- preProcess(df[numvars], method = "range")
df[numvars] <- predict(scaler, df[numvars])
print(head(df))


## Set random seed
set.seed(42)
## Create train and test indices
split <- sample.split(df, SplitRatio = 0.7)

## Create train and test dataframes
df_train <- df[split,]
df_test <- df[!split,]
# df_train <- subset(df, split == TRUE)
# df_test <- subset(df, split == FALSE)

## Print first few rows of train and test dataframes
print(head(df_train))
print(head(df_test))
print(dim(df))
print(dim(df_train))
print(dim(df_test))


corr <- cor(df_train)
print(corr)
corplot <- corrplot(corr, method = 'color')
print(corplot)

x_train <- subset(df_train, select = -target)
y_train <- as.factor(df_train$target)


ctrl <- trainControl(method="cv", number=5)
modelrf <- train(x_train, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')


# Membuat kontrol RFE
# control <- rfeControl(functions = leapsFuncs, method = "cv", number =10)
control <- rfeControl(functions = rfFuncs, method = "cv", number =10)
# Menjalankan RFE
rfe_model <- rfe(x = x_train, y = y_train, sizes = c(3), rfeControl = control)
# Mendapatkan hasil support dan ranking
support <- predictors(rfe_model)
ranking <- varImp(rfe_model, scale = FALSE)

print(support)
print(ranking)

# Membuat dataframe hasil
hasil <- data.frame(
    Features = rownames(ranking),
    Support = rownames(ranking) %in% support,
    Rank = ranking$Overall
)

print(hasil)

support <- hasil$Features[hasil$Support == TRUE & hasil$Rank > 20]
not_support <- hasil$Features[hasil$Support == FALSE | hasil$Rank < 20]
print(support)
print(not_support)

x_train_rfe <- x_train[support]
print(head(x_train_rfe))

modelrf <- train(x_train_rfe, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')


# Menghitung VIF
# Gabungkan x_train_rfe dan y_train menjadi satu data frame
data_train <- cbind(x_train_rfe, response = y_train)
# Membuat formula model linier dengan semua fitur dalam x_train_rfe
formula <- as.formula(paste("response ~", paste(names(x_train_rfe), collapse = " + ")))
# Membuat model linier menggunakan formula tersebut
model <- glm(formula, data = data_train, family = binomial)
# Menghitung VIF
vif_values <- vif(model)
print(vif_values)

# Membuat data frame dengan hasil VIF
vif_df <- data.frame(
    Features = names(vif_values),
    VIF = vif_values
)
vif_df <- vif_df[order(-vif_df$VIF), ]

print(vif_df)



y_pred <- predict(modelrf, x_train_rfe)
df_pred <- data.frame(
    y_train = y_train,
    y_pred = y_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_train != y_pred)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred, reference = y_train)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred, reference = y_train)
print(cm)

x_test <- subset(df_test, select = -target)
x_test_rfe <- x_test[support]
y_test <- as.factor(df_test$target)


y_pred <- predict(modelrf, x_test_rfe)
df_pred <- data.frame(
    y_test = y_test,
    y_pred = y_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_test != y_pred)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred, reference = y_test)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred, reference = y_test)
print(cm)


